src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala

```diff
@@ -0,0 +1,245 @@
+  val MatePrimerCoordinateTag  : String = "am"
+  /** The SAM tag to use for the current record's assigned primer identifier */
+  val AmpliconIdentifierTag    : String = "ip"
+  /** The SAM tag to use for the current record's mate's assigned primer identifier */
+  val MateAmpliconIdentifierTag: String = "im"
+
+  def tags: Seq[String] = Seq(PrimerCoordinateTag, MatePrimerCoordinateTag, AmpliconIdentifierTag, MateAmpliconIdentifierTag)
+
+}
+
+/** Labels (add SAM tags) to records based on assigned amplicons.
+  *
+  * The output will have the following tags added:
+  * - ap: the assigned primer coordinates (ex. `chr1:1010873-1010894`)
+  * - am: the mate's assigned primer coordinates (ex. `chr1:1011118-1011137`)
+  * - ip: the assigned amplicon id
+  * - im: the mate's assigned amplicon id (or `=` if the same as the assigned amplicon)
+  *
+  * If not primer/amplicon was found, then no tag will be written.
+  *
+  * @param metrics a map of amplicon to [[AssignPrimersMetric]] to collect metrics, may be empty
+  * @param primerCoordinatesTag the SAM tag to store the primer genomic coordinates for the read
+  * @param matePrimerCoordinatesTag the SAM tag to store the primer genomic coordinates for the read's mate
+  * @param ampliconIdentifierTag the SAM tag to store the amplicon identifier for the read
+  * @param mateAmpliconIdentifierTag the SAM tag to store  the amplicon identifier for the read's mate
+  */
+class AmpliconLabeller(val metrics: Map[Amplicon, AssignPrimersMetric] = Map.empty,
+                       val primerCoordinatesTag: String = AssignPrimers.PrimerCoordinateTag,
+                       val matePrimerCoordinatesTag: String = AssignPrimers.MatePrimerCoordinateTag,
+                       val ampliconIdentifierTag: String = AssignPrimers.AmpliconIdentifierTag,
+                       val mateAmpliconIdentifierTag: String = AssignPrimers.MateAmpliconIdentifierTag,
+                      ) {
+  /** Labels (adds SAM tags) to a record based on the assigned amplicon(s) for itself and potentially its mate.
+    *
+    * @param rec the record to label
+    * @param recAmplicon the assigned amplicon for the record
+    * @param mateAmplicon the assigned amplicon fo the record's mate (if paired)
+    */
+  def label(rec: SamRecord,
+            recAmplicon: Option[Amplicon] = None,
+            mateAmplicon: Option[Amplicon] = None): Unit = {
+    // Find the primer for the current read
+    recAmplicon.foreach { amp =>
+      // update metrics
+      metrics.get(amp).foreach { metric =>
+        if (rec.positiveStrand) metric.left += 1 else metric.right += 1
+        if (rec.unpaired || rec.firstOfPair) metric.r1s += 1 else metric.r2s += 1
+      }
+      // assign ap/ip
+      rec(primerCoordinatesTag)   = if (rec.positiveStrand) amp.leftPrimerString else amp.rightPrimerString
+      rec(ampliconIdentifierTag) = amp.identifier
+    }
+
+    if (rec.paired) {
```

Comment:
Do you need this?  Seems weird you'd get a `mateAmplicon` if there is no mate.

Reply:
I'll add a `require` instead